File: backend/routes/jobsDisplay.js
const express = require('express');
const router = express.Router();

// Mock data for demonstration - in real implementation, this would come from your database
let mockJobs = [
  {
    job_id: 'job_12345678',
    client_id: 'client_001',
    node_id: 'node_001',
    node_ip: '192.168.1.100',
    ssh_username: 'job_12345678',
    ssh_info: 'ssh job_12345678@192.168.1.100',
    status: 'running',
    created_at: new Date().toISOString(),
    expires_at: new Date(Date.now() + 3600000).toISOString(), // 1 hour from now
    payment_amount: 50.00
  }
];

let mockNodes = [
  {
    node_id: 'node_001',
    ip_address: '192.168.1.100',
    zerotier_ip: '10.147.17.100',
    status: 'busy',
    current_job: 'job_12345678',
    ssh_user: 'job_12345678',
    capabilities: {
      cpu_cores: 8,
      ram_gb: 32,
      gpu: 'RTX 4090',
      storage_gb: 1000
    },
    pricing: {
      hourly_rate: 2.50,
      currency: 'USD'
    },
    last_seen: new Date().toISOString()
  },
  {
    node_id: 'node_002',
    ip_address: '192.168.1.101',
    zerotier_ip: '10.147.17.101',
    status: 'active',
    current_job: null,
    ssh_user: null,
    capabilities: {
      cpu_cores: 16,
      ram_gb: 64,
      gpu: 'RTX 4080',
      storage_gb: 2000
    },
    pricing: {
      hourly_rate: 3.00,
      currency: 'USD'
    },
    last_seen: new Date().toISOString()
  }
];

// @desc    Get all active jobs with SSH access
// @route   GET /api/jobs-display/active-jobs
// @access  Public
router.get('/active-jobs', async (req, res) => {
  try {
    // In real implementation, query your database for active jobs
    const activeJobs = mockJobs.filter(job => 
      job.status === 'running' || job.status === 'accepted'
    );

    res.status(200).json({
      success: true,
      count: activeJobs.length,
      data: activeJobs
    });
  } catch (error) {
    console.error('Error fetching active jobs:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch active jobs'
    });
  }
});

// @desc    Get all rental nodes
// @route   GET /api/jobs-display/rental-nodes
// @access  Public
router.get('/rental-nodes', async (req, res) => {
  try {
    res.status(200).json({
      success: true,
      count: mockNodes.length,
      data: mockNodes
    });
  } catch (error) {
    console.error('Error fetching rental nodes:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch rental nodes'
    });
  }
});

// @desc    Get job and node statistics
// @route   GET /api/jobs-display/job-stats
// @access  Public
router.get('/job-stats', async (req, res) => {
  try {
    const totalJobs = mockJobs.length;
    const activeJobs = mockJobs.filter(job => 
      job.status === 'running' || job.status === 'accepted'
    ).length;
    
    const totalNodes = mockNodes.length;
    const availableNodes = mockNodes.filter(node => node.status === 'active').length;
    const busyNodes = mockNodes.filter(node => node.status === 'busy').length;
    
    const utilization = totalNodes > 0 ? Math.round((busyNodes / totalNodes) * 100) : 0;

    const stats = {
      jobs: {
        total: totalJobs,
        active: activeJobs,
        completed: totalJobs - activeJobs
      },
      nodes: {
        total: totalNodes,
        available: availableNodes,
        busy: busyNodes,
        utilization: utilization
      }
    };

    res.status(200).json({
      success: true,
      data: stats
    });
  } catch (error) {
    console.error('Error fetching job stats:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to fetch job statistics'
    });
  }
});

// @desc    Create SSH access for a job (when payment is confirmed)
// @route   POST /api/jobs-display/jobs/:jobId/ssh-access
// @access  Public (in real app, this should be protected)
router.post('/jobs/:jobId/ssh-access', async (req, res) => {
  try {
    const { jobId } = req.params;
    
    // Find the job
    const jobIndex = mockJobs.findIndex(job => job.job_id === jobId);
    if (jobIndex === -1) {
      return res.status(404).json({
        success: false,
        message: 'Job not found'
      });
    }

    const job = mockJobs[jobIndex];
    if (job.status !== 'accepted') {
      return res.status(400).json({
        success: false,
        message: 'Job must be accepted before creating SSH access'
      });
    }

    // Generate SSH username
    const sshUsername = `job_${jobId.slice(-8)}`;
    
    // Update job
    mockJobs[jobIndex] = {
      ...job,
      ssh_username: sshUsername,
      ssh_info: `ssh ${sshUsername}@${job.node_ip}`,
      status: 'running'
    };

    res.status(200).json({
      success: true,
      data: {
        job_id: jobId,
        ssh_username: sshUsername,
        ssh_command: `ssh ${sshUsername}@${job.node_ip}`,
        node_ip: job.node_ip,
        message: 'SSH access created successfully'
      }
    });
  } catch (error) {
    console.error('Error creating SSH access:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to create SSH access'
    });
  }
});

// @desc    Remove SSH access for a job (when job ends)
// @route   DELETE /api/jobs-display/jobs/:jobId/ssh-access
// @access  Public (in real app, this should be protected)
router.delete('/jobs/:jobId/ssh-access', async (req, res) => {
  try {
    const { jobId } = req.params;
    
    // Find the job
    const jobIndex = mockJobs.findIndex(job => job.job_id === jobId);
    if (jobIndex === -1) {
      return res.status(404).json({
        success: false,
        message: 'Job not found'
      });
    }

    const job = mockJobs[jobIndex];
    const removedSshUser = job.ssh_username;

    // Update job
    mockJobs[jobIndex] = {
      ...job,
      ssh_username: null,
      ssh_info: null,
      status: 'completed'
    };

    // Update corresponding node
    const nodeIndex = mockNodes.findIndex(node => node.node_id === job.node_id);
    if (nodeIndex !== -1) {
      mockNodes[nodeIndex] = {
        ...mockNodes[nodeIndex],
        status: 'active',
        current_job: null,
        ssh_user: null
      };
    }

    res.status(200).json({
      success: true,
      data: {
        job_id: jobId,
        removed_ssh_user: removedSshUser,
        message: 'SSH access removed successfully'
      }
    });
  } catch (error) {
    console.error('Error removing SSH access:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to remove SSH access'
    });
  }
});

// @desc    Add a test job (for demonstration)
// @route   POST /api/jobs-display/test-job
// @access  Public
router.post('/test-job', async (req, res) => {
  try {
    const newJobId = `job_${Date.now().toString().slice(-8)}`;
    const availableNode = mockNodes.find(node => node.status === 'active');
    
    if (!availableNode) {
      return res.status(400).json({
        success: false,
        message: 'No available nodes'
      });
    }

    const newJob = {
      job_id: newJobId,
      client_id: `client_${Math.random().toString(36).substr(2, 6)}`,
      node_id: availableNode.node_id,
      node_ip: availableNode.ip_address,
      ssh_username: null,
      ssh_info: null,
      status: 'accepted',
      created_at: new Date().toISOString(),
      expires_at: new Date(Date.now() + 3600000).toISOString(),
      payment_amount: availableNode.pricing.hourly_rate
    };

    mockJobs.push(newJob);

    // Update node status
    const nodeIndex = mockNodes.findIndex(node => node.node_id === availableNode.node_id);
    mockNodes[nodeIndex].status = 'busy';
    mockNodes[nodeIndex].current_job = newJobId;

    res.status(201).json({
      success: true,
      data: newJob,
      message: 'Test job created successfully'
    });
  } catch (error) {
    console.error('Error creating test job:', error);
    res.status(500).json({
      success: false,
      message: 'Failed to create test job'
    });
  }
});

module.exports = router;

File: backend/app/api/v1/endpoints/jobs_display.py
from fastapi import APIRouter, HTTPException, Depends
from sqlalchemy.orm import Session
from typing import List, Optional
from app.core.database import get_db
from app.models.models import Job, ComputeNode
from app.schemas.jobs import JobResponse, JobStatus
from app.schemas.nodes import NodeResponse, NodeStatus
from pydantic import BaseModel
from datetime import datetime

router = APIRouter()

class ActiveJobResponse(BaseModel):
    job_id: str
    client_id: str
    node_id: str
    node_ip: str
    ssh_username: Optional[str] = None
    ssh_info: Optional[str] = None
    status: str
    created_at: datetime
    expires_at: Optional[datetime] = None
    payment_amount: Optional[float] = None

class RentalNodeResponse(BaseModel):
    node_id: str
    ip_address: str
    zerotier_ip: Optional[str] = None
    status: str
    current_job: Optional[str] = None
    ssh_user: Optional[str] = None
    capabilities: dict
    pricing: dict
    last_seen: datetime

@router.get("/active-jobs", response_model=List[ActiveJobResponse])
async def get_active_jobs(db: Session = Depends(get_db)):
    """Get all currently active jobs with rental node information"""
    
    try:
        # Query active jobs with node information
        jobs = db.query(Job).filter(
            Job.status.in_([JobStatus.RUNNING, JobStatus.ACCEPTED])
        ).all()
        
        active_jobs = []
        for job in jobs:
            # Get node information
            node = db.query(ComputeNode).filter(
                ComputeNode.id == job.node_id
            ).first()
            
            job_response = ActiveJobResponse(
                job_id=str(job.id),
                client_id=str(job.client_id),
                node_id=str(job.node_id) if job.node_id else "Unknown",
                node_ip=node.ip_address if node else "Unknown",
                ssh_username=job.ssh_username if hasattr(job, 'ssh_username') else None,
                ssh_info=f"ssh {job.ssh_username}@{node.ip_address}" if (hasattr(job, 'ssh_username') and job.ssh_username and node) else None,
                status=job.status.value,
                created_at=job.created_at,
                expires_at=job.expires_at if hasattr(job, 'expires_at') else None,
                payment_amount=float(job.payment_amount) if job.payment_amount else None
            )
            active_jobs.append(job_response)
        
        return active_jobs
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve active jobs: {str(e)}")

@router.get("/rental-nodes", response_model=List[RentalNodeResponse])
async def get_rental_nodes(db: Session = Depends(get_db)):
    """Get all available rental nodes with their current status"""
    
    try:
        nodes = db.query(ComputeNode).filter(
            ComputeNode.status.in_([NodeStatus.ACTIVE, NodeStatus.BUSY])
        ).all()
        
        rental_nodes = []
        for node in nodes:
            # Check for current job
            current_job = db.query(Job).filter(
                Job.node_id == node.id,
                Job.status.in_([JobStatus.RUNNING, JobStatus.ACCEPTED])
            ).first()
            
            node_response = RentalNodeResponse(
                node_id=str(node.id),
                ip_address=node.ip_address,
                zerotier_ip=node.zerotier_ip if hasattr(node, 'zerotier_ip') else None,
                status=node.status.value,
                current_job=str(current_job.id) if current_job else None,
                ssh_user=current_job.ssh_username if (current_job and hasattr(current_job, 'ssh_username')) else None,
                capabilities=node.capabilities if node.capabilities else {},
                pricing=node.pricing if node.pricing else {},
                last_seen=node.last_seen
            )
            rental_nodes.append(node_response)
        
        return rental_nodes
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve rental nodes: {str(e)}")

@router.get("/job-stats")
async def get_job_statistics(db: Session = Depends(get_db)):
    """Get overall job and rental statistics"""
    
    try:
        total_jobs = db.query(Job).count()
        active_jobs = db.query(Job).filter(
            Job.status.in_([JobStatus.RUNNING, JobStatus.ACCEPTED])
        ).count()
        
        total_nodes = db.query(ComputeNode).count()
        active_nodes = db.query(ComputeNode).filter(
            ComputeNode.status == NodeStatus.ACTIVE
        ).count()
        busy_nodes = db.query(ComputeNode).filter(
            ComputeNode.status == NodeStatus.BUSY
        ).count()
        
        return {
            "jobs": {
                "total": total_jobs,
                "active": active_jobs,
                "completed": total_jobs - active_jobs
            },
            "nodes": {
                "total": total_nodes,
                "available": active_nodes,
                "busy": busy_nodes,
                "utilization": round((busy_nodes / total_nodes * 100) if total_nodes > 0 else 0, 2)
            }
        }
        
    except Exception as e:
        raise HTTPException(status_code=500, detail=f"Failed to retrieve statistics: {str(e)}")

@router.post("/jobs/{job_id}/ssh-access")
async def create_ssh_access(job_id: str, db: Session = Depends(get_db)):
    """Create SSH access for a job (called when payment is confirmed)"""
    
    try:
        job = db.query(Job).filter(Job.id == job_id).first()
        if not job:
            raise HTTPException(status_code=404, detail="Job not found")
        
        if job.status != JobStatus.ACCEPTED:
            raise HTTPException(status_code=400, detail="Job must be accepted before creating SSH access")
        
        # Get node information
        node = db.query(ComputeNode).filter(ComputeNode.id == job.node_id).first()
        if not node:
            raise HTTPException(status_code=404, detail="Node not found")
        
        # Generate SSH username (this would integrate with the SSH manager)
        import uuid
        ssh_username = f"job_{str(uuid.uuid4()).replace('-', '')[:8]}"
        
        # Update job with SSH information
        job.ssh_username = ssh_username
        job.status = JobStatus.RUNNING
        db.commit()
        
        return {
            "job_id": str(job.id),
            "ssh_username": ssh_username,
            "ssh_command": f"ssh {ssh_username}@{node.ip_address}",
            "node_ip": node.ip_address,
            "message": "SSH access created successfully"
        }
        
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Failed to create SSH access: {str(e)}")

@router.delete("/jobs/{job_id}/ssh-access")
async def remove_ssh_access(job_id: str, db: Session = Depends(get_db)):
    """Remove SSH access for a job (called when job ends)"""
    
    try:
        job = db.query(Job).filter(Job.id == job_id).first()
        if not job:
            raise HTTPException(status_code=404, detail="Job not found")
        
        # Remove SSH username and update status
        ssh_username = job.ssh_username if hasattr(job, 'ssh_username') else None
        job.ssh_username = None
        job.status = JobStatus.COMPLETED
        db.commit()
        
        return {
            "job_id": str(job.id),
            "removed_ssh_user": ssh_username,
            "message": "SSH access removed successfully"
        }
        
    except Exception as e:
        db.rollback()
        raise HTTPException(status_code=500, detail=f"Failed to remove SSH access: {str(e)}")
